rllib/models/specs/checker.py

```
@@ -0,0 +1,290 @@
+        ...         ...
+
+        >>> model = MyModel()
+        >>> model.forward({"obs": torch.randn(32, 64)}) # No error
+        >>> model.forward({"obs": torch.randn(32, 32)}) # raises ValueError
+
+    Args:
+        func: The instance method to decorate. It should be a callable that takes
+            `self` as the first argument, `input_data` as the second argument and any
+            other keyword argument thereafter.
+        input_spec: `self` should have an instance attribute whose name matches the
+            string in input_spec and returns the `SpecDict`, `Spec`, or simply the
+            `Type` that the `input_data` should comply with. It can alos be None or
+            empty list / dict to enforce no input spec.
+        filter: If True, and `input_data` is a nested dict the `input_data` will be
+            filtered by its corresponding spec tree structure and then passed into the
+            implemented function to make sure user is not confounded with unnecessary
+            data.
+        cache: If True, only checks the data validation for the first time the
+            instance method is called.
+
+    Returns:
+        A wrapped instance method. In case of `cache=True`, after the first invokation
+        of the decorated method, the intance will have `__checked_input_specs_cache__`
+        attribute that stores which method has been invoked at least once. This is a
+        special attribute that can be used for the cache itself. The wrapped class
+        method also has a special attribute `__checked_input_specs__` that marks the
+        method as decorated.
+    """
+
+    def decorator(func):
+        @functools.wraps(func)
+        def wrapper(self, input_data, **kwargs):
+            if cache and not hasattr(self, "__checked_input_specs_cache__"):
+                self.__checked_input_specs_cache__ = {}
+
+            input_data_ = input_data
+            if input_spec:
+                input_spec_ = getattr(self, input_spec)
```

Should be maybe throw an error here? getattr will return None and so we simply don't check anything if someone mistypes input_spec?

---

good call.

---

it can be None tho. Added a special fix for this.